### PR TITLE
Storage: Use prepared statements

### PR DESCRIPTION
**Pull request by tsvisabo**
* Use prepare statements and bind …
* Build model against 3 trits per tryte
Fixes #147

### CODE REVIEW

File: common/storage/sql/sqlite3/storage.c [hunk 1]
```diff
@@ -138,37 +170,49 @@ extern retcode_t iota_stor_load_hashes(const connection_t* const conn,
...
+  while (sqlite3_step(sqlite_statement) ==
+         SQLITE_ROW) {  // While query has result-rows.
+    buffer = sqlite3_column_blob(sqlite_statement, 0);
+    memcpy(pack->hashes[pack->num_loaded++]->trits, buffer, FLEX_TRIT_SIZE_243);
   }
 
-  return 0;
+  //  Step, Clear and Reset the statement after each bind.
+  rc = sqlite3_step(sqlite_statement);
+  rc = sqlite3_clear_bindings(sqlite_statement);
```
Comment (th0br0) [inline common/storage/sql/sqlite3/storage.c]: I think you want to do `rc |=` so the exception carries forward?

File: common/storage/sql/sqlite3/tests/test_sqlite3.c [hunk 2]
```diff
@@ -88,8 +88,8 @@ void test_stored_load_hashes_by_address(void) {
     pack.hashes[i] = trit_array_new(243);
   }
   pack.max_hashes = 5;
-  trit_array_p key = trit_array_new(243);
-  memcpy(key->trits, TEST_TRANSACTION.address, FLEX_TRIT_SIZE_243);
+  trit_array_p key = trit_array_new(81);
+  memcpy(key->trits, TEST_TRANSACTION.address, 81);
```
Comment (th0br0) [inline common/storage/sql/sqlite3/tests/test_sqlite3.c]: the 81 shouldn't be hardcoded here. use `FLEX_TRIT_SIZE_81`?

File: common/trinary/BUILD [hunk 3]
```diff
@@ -109,10 +109,7 @@ cc_library(
     srcs = ["trit_array.c"],
     hdrs = ["trit_array.h"],
     deps = [
-        ":trit_byte",
-        ":trit_tryte",
-        ":trits",
-        "//common:stdint",
+        ":trit_array_3",
```
Comment (th0br0) [inline common/trinary/BUILD:110]: While you're at it, remove `srcs` and `hdrs`? :)

File: common/storage/sql/sqlite3/storage.c [hunk 4]
```diff
@@ -33,102 +32,167 @@ retcode_t iota_stor_destroy() {
...
+  sqlite3_bind_blob(sqlite_statement, 9, data_in->hash, FLEX_TRIT_SIZE_243,
+                    NULL);
+  sqlite3_step(sqlite_statement);
+  if (rc != SQLITE_OK) {
+    log_error(SQLITE3_LOGGER_ID,
+              "Failed in step, sqlite3 code is: %\" PRIu64 \"", rc);
+    return RC_SQLITE3_FAILED_STEP;
+  }
 
+  rc = sqlite3_finalize(sqlite_statement);  //  Finalize the prepared statement.
```
Comment (th0br0) [inline common/storage/sql/sqlite3/storage.c]: sqlite3 prepared statements are reusable (see `sqlite3_clear_bindings`) but are *not* thread-safe.

Can you add an issue for adding the prepared statements to the connection handle?
However, we'll still need to figure out how to guarantee thread-safety in this case (TLS?)

File: common/storage/sql/sqlite3/storage.c [hunk 5]
```diff
@@ -33,102 +32,167 @@ retcode_t iota_stor_destroy() {
...
 retcode_t iota_stor_store(const connection_t* const conn,
                           const iota_transaction_t data_in) {
   char statement[MAX_STORE_STATEMENT_SIZE];
   iota_transactions_insert_statement(data_in, statement,
                                      MAX_STORE_STATEMENT_SIZE);
-
   char* err_msg = 0;
-  int rc = sqlite3_exec((sqlite3*)conn->db, statement, 0, 0, &err_msg);
+  sqlite3_stmt* sqlite_statement = 0;
+  char* buffer;
```
Comment (th0br0) [inline common/storage/sql/sqlite3/storage.c]: not used?

File: common/storage/sql/sqlite3/storage.c [hunk 6]
```diff
@@ -33,102 +32,167 @@ retcode_t iota_stor_destroy() {
...
   char statement[MAX_SELECT_STATEMENT_SIZE];
-  iota_transactions_select_statement(col_name, key, statement,
+  iota_transactions_select_statement(col_name, statement,
                                      MAX_SELECT_STATEMENT_SIZE);
 
   char* err_msg = 0;
-  int rc = sqlite3_exec((sqlite3*)conn->db, statement, select_transactions_cb,
-                        (void*)pack, &err_msg);
+  sqlite3_stmt* sqlite_statement = 0;
+  char* buffer;
```
Comment (th0br0) [inline common/storage/sql/sqlite3/storage.c:130]: not used?

File: common/storage/sql/sqlite3/storage.c [hunk 7]
```diff
@@ -33,102 +32,167 @@ retcode_t iota_stor_destroy() {
...
 retcode_t iota_stor_exist(const connection_t* const conn,
                           const char* index_name, const trit_array_p key,
                           bool* exist) {
-  char statement[MAX_EXIST_STATEMENT_SIZE];
-  iota_transactions_exist_statement(index_name, key, statement,
-                                    MAX_EXIST_STATEMENT_SIZE);
-
   char* err_msg = 0;
+  sqlite3_stmt* sqlite_statement = 0;
+  char* buffer;
```
Comment (th0br0) [inline common/storage/sql/sqlite3/storage.c]: not used?

**Review by th0br0 — CHANGES_REQUESTED** [review]

File: common/storage/sql/sqlite3/storage.c (hunk 4 above)
Comment (tsvisabo) [reply common/storage/sql/sqlite3/storage.c]: sure

File: common/trinary/BUILD (hunk 3 above)
Comment (tsvisabo) [reply common/trinary/BUILD:110]: done

File: common/storage/sql/sqlite3/storage.c [hunk 8]
```diff
@@ -79,23 +78,29 @@ retcode_t iota_stor_store(const connection_t* const conn,
...
+  rc |= sqlite3_bind_blob(sqlite_statement, 8, data_in->nonce,
+                          FLEX_TRIT_SIZE_81, NULL);
+  rc |= sqlite3_bind_blob(sqlite_statement, 9, data_in->hash,
+                          FLEX_TRIT_SIZE_243, NULL);
+  if (rc != SQLITE_OK) {
+    log_error(SQLITE3_LOGGER_ID,
+              "Failed in binnding, sqlite3 code is: %\" PRIu64 \"", rc);
+    return RC_SQLITE3_FAILED_BINDING;
+  }
   sqlite3_step(sqlite_statement);
```
Comment (th0br0) [inline common/storage/sql/sqlite3/storage.c]: you're missing a `rc =` here

File: common/storage/sql/sqlite3/storage.c [hunk 9]
```diff
@@ -33,102 +32,171 @@ retcode_t iota_stor_destroy() {
...
+  rc |= sqlite3_bind_blob(sqlite_statement, 8, data_in->nonce,
+                          FLEX_TRIT_SIZE_81, NULL);
+  rc |= sqlite3_bind_blob(sqlite_statement, 9, data_in->hash,
+                          FLEX_TRIT_SIZE_243, NULL);
+  if (rc != SQLITE_OK) {
+    log_error(SQLITE3_LOGGER_ID,
+              "Failed in binnding, sqlite3 code is: %\" PRIu64 \"", rc);
+    return RC_SQLITE3_FAILED_BINDING;
+  }
+  sqlite3_step(sqlite_statement);
```
Comment (th0br0) [inline common/storage/sql/sqlite3/storage.c]: missing `rc =` here

File: common/storage/sql/sqlite3/storage.c [hunk 10]
```diff
@@ -33,102 +32,171 @@ retcode_t iota_stor_destroy() {
...
   if (rc != SQLITE_OK) {
-    log_error(SQLITE3_LOGGER_ID, "Failed in statement", statement);
-    sqlite3_free(err_msg);
-    return RC_SQLITE3_FAILED_INSERT_DB;
+    log_error(SQLITE3_LOGGER_ID,
```
Comment (th0br0) [inline common/storage/sql/sqlite3/storage.c:106]: I think you want to test `rc != SQLITE_DONE` ?

**Review by th0br0 — CHANGES_REQUESTED** [review]

File: common/storage/sql/sqlite3/storage.c [hunk 11]
```diff
@@ -192,6 +205,11 @@ retcode_t iota_stor_exist(const connection_t* const conn,
 
   rc = sqlite3_finalize(sqlite_statement);  //  Finalize the prepared statement.
   if (rc != SQLITE_OK) {
+    sqlite3_finalize(sqlite_statement);  //  Finalize the prepared statement.
```
Comment (th0br0) [inline common/storage/sql/sqlite3/storage.c]: not needed; you're already finalizing before this call

File: common/storage/sql/sqlite3/storage.c [hunk 12]
```diff
@@ -285,6 +317,9 @@ retcode_t iota_stor_load_hashes_approvers(const connection_t* const conn,
...
```
Comment (th0br0) [inline common/storage/sql/sqlite3/storage.c]: duplicate finalize

**Review by th0br0 — CHANGES_REQUESTED** [review]

**Review by th0br0 — APPROVED** [review]

**Pull request merged**